compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -1259,6 +1259,17 @@ pub(super) fn check_transparent<'tcx>(tcx: TyCtxt<'tcx>, adt: ty::AdtDef<'tcx>)
                 ty::Tuple(list) => list.iter().try_for_each(|t| check_non_exhaustive(tcx, t)),
                 ty::Array(ty, _) => check_non_exhaustive(tcx, *ty),
                 ty::Adt(def, args) => {
+                    if matches!(
+                        tcx.as_lang_item(def.did()),
+                        Some(LangItem::UnsafeCell | LangItem::ManuallyDrop | LangItem::MaybeUninit)
+                    ) || matches!(
+                        tcx.get_diagnostic_name(def.did()),
```

Comment:
I don't think diagnostic items should be used in a way that *affects* whether something is linted. They should only be used to change the quality of a diagnostic (i.e. its labels or its message). These should be turned into lang items.

Reply:
Also, is there a reason why we don't apply this to any `repr(transparent)` ADT?

Reply:
> Also, is there a reason why we don't apply this to any `repr(transparent)` ADT?

As mentioned in the original issue

> More generally, quoting [the nomicon](https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent):
> 
> > This repr is only considered part of the public ABI of a type if either the single field is pub, or if its layout is documented in prose. Otherwise, the layout should not be relied upon by other crates.
> 
> So, in `ManuallyDrop`'s case, the single field is not pub, but the layout is documented in prose and so should be considered part of the ABI. Is it possible there are other types out there with such a layout guarantee in prose, which might also be incorrectly flagged by this lint?

Reply:
> These should be turned into lang items.

Is there some sort of FCP required for this?

Reply:
No, you just need to do something like this: f4f57bfccb98522f4aaa2942baaa5a381c7d952d

Reply:
For `Pin` we also [guarantee the layout in prose](https://doc.rust-lang.org/nightly/core/pin/struct.Pin.html#layout-and-abi).

I think it would be better to introduce an internal attribute (e.g. `rustc_pub_transparent`) to mark those types rather than special case them all individually just for this lint.

Reply:
Yep, I think a builtin attr would be nice.